### PR TITLE
Implements `CertificateRequest` for client authentication

### PR DESCRIPTION
**Pull request by bkchr**
The context has a new parameter `client_authentication` to
enable/disable the feature.

Should fix: https://github.com/h2o/picotls/issues/116

I create this pull request to get some feedback. The implementation should be finished.

I still want to add some testing to ensure that everything works as expected :)

### CODE REVIEW

**Comment by bkchr** [comment]
@kazuho Now I also added a test that checks the full handshake. If you want me to add more tests, I'm open, just say with combinations should be tested :) 
I also fixed a lot of bugs (in my new code) today, will trying to get my test working.

I also added a lot of braces after if statements, to prevent bugs in the future (think of apple and goto ;)), I hope that is okay.

**Comment by bkchr** [comment]
Hmm clang still fails :( I will need to investigate that!

**Comment by kazuho** [comment]
Thank you for the PR. You are really quick!

> I also added a lot of braces after if statements, to prevent bugs in the future (think of apple and goto ;)), I hope that is okay.

I would appreciate it if you could revert the changes irrelevant to client authentication. The review becomes so hard, makes harder to use `git blame` in the future.

**Comment by bkchr** [comment]
Okay, I will revert them. Are you open for a new pull request, that takes care of these braces?

**Comment by bkchr** [comment]
@kazuho done :) I also fixed the clang bug, at least it works locally :)

File: lib/picotls.c [hunk 1]
```diff
@@ -1849,6 +2044,17 @@ static int client_handle_finished(ptls_t *tls, ptls_buffer_t *sendbuf, ptls_iove
         }
         if ((ret = retire_early_data_secret(tls, 1)) != 0)
             goto Exit;
+    } else if (tls->client.certificate_request_received == 1) {
```
Comment (kazuho) [inline lib/picotls.c]: Is use of early-data and client-authentication exclusive?

If not, I think `else` should be removed.

Comment (bkchr) [reply lib/picotls.c]: Based on the following figure: https://tlswg.github.io/tls13-spec/draft-ietf-tls-tls13.html#zero-rtt-data
I would assume that they are exclusive and the `else` is required. As the client is required to have the `PSK` of the server, it already has to had a connection once to the server.

File: include/picotls.h [hunk 2]
```diff
@@ -462,6 +462,11 @@ struct st_ptls_context_t {
      *
      */
     ptls_update_open_count_t *update_open_count;
+    /**
+     * if set, the server requests client certificates
+     * to authenticate the client.
+     */
+    unsigned client_authentication : 1;
```
[the remark this answers is not in the document]
Comment (bkchr) [reply include/picotls.h]: Yeah, good idea with the bit fields.

**Comment by bkchr** [comment]
@kazuho Integrated your feedback :) Please also look again at my explanation for the `else if`.

**Comment by kazuho** [comment]
@bkchr Thank you for your changes!

> I would assume that they are exclusive and the `else` is required. As the client is required to have the `PSK` of the server, it already has to had a connection once to the server.

The general rule will be that if there's a prohibited action, we need to make sure that we do not make such action and also emit a handshake failure when such action is taken by peer.

In this particular case, I have gone back to the draft, and found the following statement, which confirms your understanding that the two action cannot be done together.
> Servers which are authenticating with a PSK MUST NOT send the CertificateRequest message in the main handshake, though they MAY send it in post-handshake authentication (see Section 4.6.2) provided that the client has sent the “post_handshake_auth” extension (see Section 4.2.6).
> https://tlswg.github.io/tls13-spec/draft-ietf-tls-tls13.html#rfc.section.4.3.2

So I think we need to make sure that:
* picotls acting as a client aborts the handshake with an `illegal_parameter` alert (I think that is the correct error code here) when the server sends a Certificate Request extension on a resuming handshake
* picotls acting as a sever always uses full-handshake when sending a Certificate Request extension

Is my understanding correct that the first point is not implemented yet? If that is the case, I would appreciate it if you could work on the issue.

For the second point, I now see that the PR is sending Certificate Request extension only when running a full handshake. That makes me wonder how we will handle session resumption when client authentication is involved. Do you have any specific idea? Or are you simply not interested in resuming a session when client authentication is involved? If it is the latter, how about enforcing full handshake as a safeguard when `require_client_authentication` flag is set?

**Comment by bkchr** [comment]
Yeah, I could work on the first point. That should not be that complicated. 

Regarding client authentication on resuming session, as far as I understand the draft, the client authentication can also be send after `ClientFinish`. 

> Servers which are authenticating with a PSK MUST NOT send the CertificateRequest message in 
the main handshake, though they MAY send it in post-handshake authentication (see Section 4.6.2) provided that the client has sent the “post_handshake_auth” extension (see Section 4.2.6).

I think you are right, we should enforce full handshaking, when `require_client_authentication` is set and the client did not send the `post_handshake_auth` extension. A first implementation should maybe always enforce full handshaking and after we support post handshake authentication, we could switch easily for checking the extension.

**Comment by kazuho** [comment]
Thank you for the prompt answer.

> Yeah, I could work on the first point. That should not be that complicated.

I would appreciate that.

> A first implementation should maybe always enforce full handshaking and after we support post handshake authentication, we could switch easily for checking the extension.

SGTM. Let's start with MVP.

**Comment by bkchr** [comment]
@kazuho Sorry for the delay, but now I integrated all your feedback :)

**Comment by bkchr** [comment]
@kazuho ping?

File: lib/picotls.c [hunk 3]
```diff
@@ -1769,19 +1788,150 @@ static int client_handle_encrypted_extensions(ptls_t *tls, ptls_iovec_t message,
...
+    uint16_t exttype = 0;
+
+    /* certificate request context */
+    ptls_decode_open_block(src, end, 1, {
+            if (end - src > 255) {
+                ret = PTLS_ALERT_DECODE_ERROR;
+                goto Exit;
+            }
+
+            cr->certificate_request_context = ptls_iovec_init(src, end - src);
```
Comment (kazuho) [inline lib/picotls.c]: I am not sure if _not_ copying the context is safe here.

This function is called when processing a CertificateRequest message. We will be using the context when we receive ClientFinished. picotls releases memory that contained the message, once the processing of either the message or the record that contained the message finishes.

I think that that means that if the two messages arrive as different TLS records, we would be referring to a freed memory when we try to the information.

Comment (bkchr) [reply lib/picotls.c]: Yeah, you are right! I will fix that immediately :)

**Comment by bkchr** [comment]
@kazuho done :)

**Comment by bkchr** [comment]
@kazuho ping again. I had to rebase again to resolve conflicts. It would be nice if we finally get this to a point where it can be merged :)

**Comment by bkchr** [comment]
@kazuho ping.

**Comment by kazuho** [comment]
Sorry for the delay.

The PR has been merged in the commit above. I have added command line options to support client authenication as well as fixing some bugs.

**Comment by bkchr** [comment]
Ty :)

**Pull request closed without merging**